internal/auth/login_handler.go

```diff
@@ -0,0 +1,362 @@
+		panic(err)
+	}
+
+	baseClientURL := nodeConfig.ExternalURL()
+	oauthClient, err := NewOAuthClient(
+		baseClientURL+"/habitat/api/client-metadata.json", /*clientId*/
+		baseClientURL, /*clientUri*/
+		baseClientURL+"/habitat/api/auth-callback", /*redirectUri*/
+		jwk, /*secretJwk*/
+	)
+	if err != nil {
+		log.Error().Err(err).Msg("error creating oauth client")
+		return nil, err
+	}
+
+	return []api.Route{
+		&loginHandler{
+			pdsURL:            nodeConfig.InternalPDSURL(),
+			oauthClient:       oauthClient,
+			sessionStore:      sessionStore,
+			habitatNodeDomain: nodeConfig.Domain(),
+		}, &metadataHandler{
+			oauthClient: oauthClient,
+		}, &callbackHandler{
+			oauthClient:  oauthClient,
+			sessionStore: sessionStore,
+		},
+	}, nil
+}
+
+// Method implements api.Route.
+func (l *loginHandler) Method() string {
+	return http.MethodGet
+}
+
+// Pattern implements api.Route.
+func (l *loginHandler) Pattern() string {
+	return "/login"
+}
+
+// ServeHTTP implements api.Route.
+func (l *loginHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	err := r.ParseForm()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	handle := r.Form.Get("handle")
+
+	atid, err := syntax.ParseAtIdentifier(handle)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	id, loginHint, err := l.resolveIdentity(r.Context(), atid, handle)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	pdsService, ok := id.Services["atproto_pds"]
+	if !ok {
+		http.Error(w, "no pds url in identity", http.StatusInternalServerError)
+		return
+	}
+
+	dpopSession, err := newCookieSession(r, l.sessionStore, id, pdsService.URL)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// Get the key from the session
+	key, ok, err := dpopSession.GetDpopKey()
+	if !ok {
+		http.Error(w, "no key in session", http.StatusInternalServerError)
+		return
+	}
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	dpopClient := NewDpopHttpClient(key, dpopSession)
+
+	redirect, state, err := l.oauthClient.Authorize(dpopClient, id, loginHint)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	stateJson, err := json.Marshal(state)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	session, _ := l.sessionStore.New(r, "auth-session")
+	session.AddFlash(stateJson)
+
+	dpopSession.Save(r, w)
+	err = session.Save(r, w)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	http.Redirect(w, r, redirect, http.StatusSeeOther)
+}
+
+func (l *loginHandler) resolveIdentity(ctx context.Context, atID *syntax.AtIdentifier, handle string) (*identity.Identity, *string, error) {
+	pdsOnHabitat := doesHandleBelongToDomain(handle, l.habitatNodeDomain)
+	if pdsOnHabitat {
```

Comment:
up to you but now that we know pds won't be on habitat we can clean this up now. otherwise fine to clean up later